clang-tools-extra/clang-tidy/misc/BoolBitwiseOperationCheck.cpp

```diff
@@ -0,0 +1,219 @@
+                                   const LangOptions &LangOpts) {
+  if (!E)
+    return false;
+
+  const SourceLocation Start = E->getBeginLoc();
+  const SourceLocation End = E->getEndLoc();
+
+  if (Start.isMacroID() || End.isMacroID() || Start.isInvalid() ||
+      End.isInvalid())
+    return false;
+
+  const std::optional<Token> PrevTok =
+      Lexer::findPreviousToken(Start, SM, LangOpts, /*IncludeComments=*/false);
+  const std::optional<Token> NextTok =
+      Lexer::findNextToken(End, SM, LangOpts, /*IncludeComments=*/false);
+
+  return (PrevTok && PrevTok->is(tok::l_paren)) &&
+         (NextTok && NextTok->is(tok::r_paren));
+}
+
+constexpr std::array<std::pair<llvm::StringRef, llvm::StringRef>, 8U>
+    OperatorsTransformation{{{"|", "||"},
+                             {"|=", "||"},
+                             {"&", "&&"},
+                             {"&=", "&&"},
+                             {"bitand", "and"},
+                             {"and_eq", "and"},
+                             {"bitor", "or"},
+                             {"or_eq", "or"}}};
+
+static std::string translate(llvm::StringRef Value) {
+  for (const auto &[Bitwise, Logical] : OperatorsTransformation) {
+    if (Value == Bitwise)
+      return Logical.str();
+  }
+
+  return {};
+}
+
+BoolBitwiseOperationCheck::BoolBitwiseOperationCheck(StringRef Name,
+                                                     ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context),
+      StrictMode(Options.get("StrictMode", false)),
+      IgnoreMacros(Options.get("IgnoreMacros", false)) {}
+
+void BoolBitwiseOperationCheck::storeOptions(
+    ClangTidyOptions::OptionMap &Opts) {
+  Options.store(Opts, "StrictMode", StrictMode);
+  Options.store(Opts, "IgnoreMacros", IgnoreMacros);
+}
+
+void BoolBitwiseOperationCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      binaryOperator(
+          unless(isExpansionInSystemHeader()),
+          hasAnyOperatorName("|", "&", "|=", "&="),
+          hasEitherOperand(expr(hasType(booleanType()))),
+          optionally(hasParent( // to simple implement transformations like
+                                // `a&&b|c` -> `a&&(b||c)`
+              binaryOperator().bind("p"))))
+          .bind("op"),
+      this);
+}
+
+void BoolBitwiseOperationCheck::check(const MatchFinder::MatchResult &Result) {
+  const auto *MatchedExpr = Result.Nodes.getNodeAs<BinaryOperator>("op");
+
+  auto DiagEmitterProc = [MatchedExpr, this] {
+    const NamedDecl *ND = getLHSNamedDeclIfCompoundAssign(MatchedExpr);
+    return diag(MatchedExpr->getOperatorLoc(),
+                "use logical operator '%0' for boolean %select{variable "
+                "%2|values}1 instead of bitwise operator '%3'")
+           << translate(MatchedExpr->getOpcodeStr()) << (ND == nullptr) << ND
+           << MatchedExpr->getOpcodeStr();
+  };
+  auto DiagEmitter = llvm::make_scope_exit(DiagEmitterProc);
```

Comment:
I am not sure whether it is a good code pattern here to use raii for diag. maybe using lambda directly is more straight forward.

Reply:
It will have a problem when an exception will be trown from the lambda, so I will change it to use lambda directly